src/main/cpp/src/cast_string.cu

```diff
@@ -202,42 +662,14 @@ struct string_to_integer_impl {
     auto col = std::make_unique<column>(
       data_type{type_to_id<T>()}, string_col.size(), data.release(), null_mask.release());
 
-    if (ansi_mode) {
-      auto const num_nulls      = col->null_count();
-      auto const incoming_nulls = string_col.null_count();
-      auto const num_errors     = num_nulls - incoming_nulls;
-      if (num_errors > 0) {
-        auto const first_error =
-          thrust::find_if(rmm::exec_policy(stream),
-                          thrust::make_counting_iterator(0),
-                          thrust::make_counting_iterator(col->size()),
-                          row_valid_fn{col->view().null_mask(), string_col.null_mask()});
-
-        offset_type string_bounds[2];
-        cudaMemcpyAsync(&string_bounds,
-                        &string_col.offsets().data<offset_type>()[*first_error],
-                        sizeof(offset_type) * 2,
-                        cudaMemcpyDeviceToHost,
-                        stream.value());
-        stream.synchronize();
-
-        std::string dest;
-        dest.resize(string_bounds[1] - string_bounds[0]);
-
-        cudaMemcpyAsync(dest.data(),
-                        &string_col.chars().data<char const>()[string_bounds[0]],
-                        string_bounds[1] - string_bounds[0],
-                        cudaMemcpyDeviceToHost,
-                        stream.value());
-        stream.synchronize();
-
-        throw cast_error(*first_error, dest);
-      }
-    }
+    if (ansi_mode) { validate_ansi_column(col->view(), string_col, stream); }
```

Comment:
Can we validate while casting, instead of calling to a separate kernel? We can make `ansi_mode` a template parameter to reduce performance impact.

Reply:
I was torn on this myself, but I assume the majority of cases will be non-ansi and as a result, I didn't want to impact performance of the common path. If we have an ansi mode request and there are errors, I'm not concerned about the kernel launch for it. It does introduce possibly 2 kernel launches to determine if we have errors though when we count the nulls, but I chalked that up as the cost of admission.

If we pushed this content into the kernel itself, we would lose some of the commonality for it. Each thread would need to either atomic set a bit somewhere or atomic increment or we each row would have a mask similar to the null mask. The problem there is the extra memory required over the common case. This was the least-intrusive way to address this I could find.